### PR TITLE
HHH-18033 Fix LimitHandler detect wrong statement end if sql contains quoted semicolon

### PR DESCRIPTION
**Pull request by quaff**
https://hibernate.atlassian.net/browse/HHH-18033

### CODE REVIEW

**Comment by quaff** [comment]
@gavinking Please review.

File: hibernate-community-dialects/src/test/java/org/hibernate/community/dialect/RowsLimitHandlerTest.java [hunk 1]
```diff
@@ -0,0 +1,27 @@
+/*
+ * Hibernate, Relational Persistence for Idiomatic Java
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later
+ * See the lgpl.txt file in the root directory or http://www.gnu.org/licenses/lgpl-2.1.html
+ */
+package org.hibernate.community.dialect;
```
Comment (gavinking) [inline hibernate-community-dialects/src/test/java/org/hibernate/community/dialect/RowsLimitHandlerTest.java:7]: Shouldn't this be in a test package?

File: hibernate-community-dialects/src/main/java/org/hibernate/community/dialect/pagination/IngresLimitHandler.java [hunk 2]
```diff
@@ -66,7 +66,7 @@ protected boolean renderOffsetRowsKeyword() {
 	};
 
 	private static final Pattern WITH_OPTION_PATTERN =
-			Pattern.compile("\\s+with\\s+(" + String.join("|", WITH_OPTIONS) + ")\\b|\\s*(;|$)");
+			Pattern.compile("\\s+with\\s+(" + String.join("|", WITH_OPTIONS) + ")\\b|\\s*;?$");
```
Comment (gavinking) [inline hibernate-community-dialects/src/main/java/org/hibernate/community/dialect/pagination/IngresLimitHandler.java:69]: Would probably be beset to accept whitespace after the semicolon, I suppose.

**Review by gavinking — CHANGES_REQUESTED** [review]
It looks fine, just a couple of minor comments.

**Comment by gavinking** [comment]
I wonder if, since we're messing with this, we should also recognize line-ending comments of form `-- text`.

File: hibernate-community-dialects/src/test/java/org/hibernate/community/dialect/RowsLimitHandlerTest.java (hunk 1 above)
Comment (quaff) [reply hibernate-community-dialects/src/test/java/org/hibernate/community/dialect/RowsLimitHandlerTest.java:7]: There are other tests already exist in this package, I think it's better to keep it.

File: hibernate-community-dialects/src/main/java/org/hibernate/community/dialect/pagination/IngresLimitHandler.java (hunk 2 above)
Comment (quaff) [reply hibernate-community-dialects/src/main/java/org/hibernate/community/dialect/pagination/IngresLimitHandler.java:69]: I think the sql should be trimmed before reaching here.

**Comment by gavinking** [comment]
> I don't think people will pass such sql to `entityManager.createNativeQuery()`.

Probably not, but they definitely might if it's a named query, especially one declared in XML.

File: hibernate-community-dialects/src/main/java/org/hibernate/community/dialect/pagination/IngresLimitHandler.java (hunk 2 above)
Comment (gavinking) [reply hibernate-community-dialects/src/main/java/org/hibernate/community/dialect/pagination/IngresLimitHandler.java:69]: Oh OK, then an alternative would be: `|\\s*;|$`, which is perhaps a little clearer .... maybe?

File: hibernate-community-dialects/src/test/java/org/hibernate/community/dialect/RowsLimitHandlerTest.java (hunk 1 above)
Comment (gavinking) [reply hibernate-community-dialects/src/test/java/org/hibernate/community/dialect/RowsLimitHandlerTest.java:7]: Oh, well. That doesn't seem quite right to me.

@beikov is there a good reason the tests for the community dialects are in the same package as the dialect classes? Shouldn't they be somewhere like `org.hibernate.community.dialect.test` or whatever?

Comment (beikov) [reply hibernate-community-dialects/src/test/java/org/hibernate/community/dialect/RowsLimitHandlerTest.java:7]: Having tests in the same package as the Dialect under test allows to test package-private methods, which I suppose was the original reason for doing this. When we moved the classes to the community dialects module, we didn't change the package for tests apparently. Either is fine IMO, though if you want to change it, I think I'd prefer the package you are proposing.

File: hibernate-community-dialects/src/main/java/org/hibernate/community/dialect/pagination/IngresLimitHandler.java (hunk 2 above)
Comment (quaff) [reply hibernate-community-dialects/src/main/java/org/hibernate/community/dialect/pagination/IngresLimitHandler.java:69]: `\\s*;|$` will match quoted semicolon also, perhaps you mean `\\s*;$|$`?

**Comment by quaff** [comment]
> > I don't think people will pass such sql to `entityManager.createNativeQuery()`.
> 
> Probably not, but they definitely might if it's a named query, especially one declared in XML.

Same as spaces, we should strip comments before reaching here, and it should be handled in another PR.

File: hibernate-community-dialects/src/main/java/org/hibernate/community/dialect/pagination/IngresLimitHandler.java (hunk 2 above)
Comment (gavinking) [reply hibernate-community-dialects/src/main/java/org/hibernate/community/dialect/pagination/IngresLimitHandler.java:69]: Oh, right, sure, of course, I forgot what you were trying to achieve here.

Comment (gavinking) [reply hibernate-community-dialects/src/main/java/org/hibernate/community/dialect/pagination/IngresLimitHandler.java:69]: > I think the sql should be trimmed before reaching here.

Wait, are you asserting that it *does* get trimmed, or are you saying you think it *should*, but that it actually doesn't?

Comment (gavinking) [reply hibernate-community-dialects/src/main/java/org/hibernate/community/dialect/pagination/IngresLimitHandler.java:69]: > I'm not sure it actually does get trimmed or not.
> I'm asserting that it's trimmed before reaching here.

I don't understand. Those two sentences seem to contradict each other.

Comment (quaff) [reply hibernate-community-dialects/src/main/java/org/hibernate/community/dialect/pagination/IngresLimitHandler.java:69]: https://github.com/hibernate/hibernate-orm/pull/8270/files#diff-82086e5bcedd2e6969ed098a04c054677391921bb15ad6fc4abfe4cc8784f85eL91

SQL get trimmed in the only caller `DeferredResultSetAccess`, is it appropriate? and I added a dedicated tests in `QueryTests` to guard that.

Comment (gavinking) [reply hibernate-community-dialects/src/main/java/org/hibernate/community/dialect/pagination/IngresLimitHandler.java:69]: I mean.... sure, I suppose that's fine, but I don't understand why you didn't just go with my initial suggestion to use `|\\s*;?\\s*;$` or whatever. It's simpler and more self-contained, and so much less-likely to be break in future.

I don't see why this fix should be impacting code in `DeferredResultSetAccess`, which is rather a long way away.

**Comment by gavinking** [comment]
> I wonder if, since we're messing with this, we should also recognize line-ending comments of form `-- text`.

Thinking this through, it's probably impossible to distinguish a line-ending comment from a quoted string literal containing the character sequence `--` without a proper scanner.

So forget this suggestion.

File: hibernate-community-dialects/src/main/java/org/hibernate/community/dialect/pagination/IngresLimitHandler.java (hunk 2 above)
Comment (quaff) [reply hibernate-community-dialects/src/main/java/org/hibernate/community/dialect/pagination/IngresLimitHandler.java]: Amended as you suggested.

**Comment by quaff** [comment]
`Oracle12LimitHandler` fixed also.

**Review by gavinking — APPROVED** [review]
OK, great, thanks. Looks good now.

Appreciate your patience with my pickiness.

File: hibernate-core/src/test/java/org/hibernate/orm/test/dialect/AbstractLimitHandlerTest.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+		String expected = "select * from Person p where p.name like ';'" + getLimitClause() + ";";
+		assertGenerateExpectedSql(expected, sql);
+
+		sql = "select * from Person p where p.name like ';' ; ";
+		expected = "select * from Person p where p.name like ';'" + getLimitClause() + " ; ";
+		assertGenerateExpectedSql(expected, sql);
+	}
+
+	protected void assertGenerateExpectedSql(String expected, String sql) {
+		assertEquals(expected, getLimitHandler().processSql(sql, new Limit(0, 10), QueryOptions.NONE));
```
Comment (beikov) [inline hibernate-core/src/test/java/org/hibernate/orm/test/dialect/AbstractLimitHandlerTest.java:47]:
```suggestion
		assertEquals(expected, getLimitHandler().processSql(sql, getLimit(), QueryOptions.NONE));
```

**Review by beikov — APPROVED** [review]

File: hibernate-core/src/test/java/org/hibernate/orm/test/dialect/AbstractLimitHandlerTest.java (hunk 3 above)
Comment (quaff) [reply hibernate-core/src/test/java/org/hibernate/orm/test/dialect/AbstractLimitHandlerTest.java]: Resolved.

**Pull request merged**